posthog/temporal/batch_exports/spmc.py

```diff
@@ -453,27 +456,29 @@ def consumer_done_callback(task: asyncio.Task):
 
     await consumer.logger.adebug("Starting record batch consumer")
 
-    consumer_task = consumer.create_consumer_task(
-        queue=queue,
-        producer_task=producer_task,
-        max_bytes=max_bytes,
-        schema=schema,
-        json_columns=json_columns,
-        multiple_files=multiple_files,
-        include_inserted_at=include_inserted_at,
-        max_file_size_bytes=max_file_size_bytes,
-        **writer_file_kwargs or {},
-    )
-    consumer_tasks_pending.add(consumer_task)
-    consumer_task.add_done_callback(consumer_done_callback)
-
-    await asyncio.wait(consumer_tasks_pending)
-
-    if consumer_task.done():
-        consumer_task_exception = consumer_task.exception()
-
-        if consumer_task_exception is not None:
-            raise consumer_task_exception
+    # We use a TaskGroup to ensure that if the activity is cancelled, this is propagated to all pending tasks.
+    try:
+        async with asyncio.TaskGroup() as tg:
+            consumer_task = consumer.create_consumer_task(
+                tg=tg,
+                queue=queue,
+                producer_task=producer_task,
+                max_bytes=max_bytes,
+                schema=schema,
+                json_columns=json_columns,
+                multiple_files=multiple_files,
+                include_inserted_at=include_inserted_at,
+                max_file_size_bytes=max_file_size_bytes,
+                **writer_file_kwargs or {},
+            )
+            consumer_task.add_done_callback(consumer_done_callback)
+            consumer_tasks_pending.add(consumer_task)
+    except Exception:
```

Comment:
Shouldn't we be catching an exception group here?

Reply:
Otherwise I think the logic is fine: We currently only run one task, so just re-raising whatever that task threw is fine. Later we will need to handle multiple exceptions though.

Reply:
Yeah I did consider that but since we're only running a single task I thought it was simplest to keep the logic as is.

Catching an exception should catch the exception group. From [the docs](https://docs.python.org/3/library/exceptions.html#ExceptionGroup)

> This design is so that except Exception catches an `ExceptionGroup` but not `BaseExceptionGroup`.

I have also tested locally that an Exception raised in the activity gets caught in the usual way.

Reply:
Cool beans, I think we can iterate on this design as we add more consumers for performance reasons. Anyways, amazing work :+1: